include/simics/dmllib.h

```diff
@@ -1780,6 +1783,124 @@ _get_port_saved_variable(lang_void *saved_access, conf_object_t *_portobj,
+_serialize_array(const void *data, size_t elem_size,
+                 const uint32 *dimsizes, uint32 dims,
+                 _serializer_t serialize_elem) {
+    ASSERT(dims > 0);
+    size_t total_elem_count = 1;
+    for (uint32 i = 0; i < dims; ++i) {
+        total_elem_count *= dimsizes[i];
+    }
+    return _serialize_array_aux((const uint8 *)data, elem_size, dimsizes, dims,
+                                total_elem_count, serialize_elem);
+}
+
+
+static set_error_t
+_deserialize_array_aux(attr_value_t val, uint8 *data, size_t elem_size,
+                       const uint32 *dimsizes, uint32 dims,
+                       size_t total_elem_count,
+                       _deserializer_t deserialize_elem,
+                       bool elems_are_bytes) {
+    uint32 len = *dimsizes;
+
+    // Allow the final dimension to be represented as data if elems_are_bytes
+    bool data_allowed = elems_are_bytes && dims == 1;
+    if (data_allowed && SIM_attr_is_data(val)) {
+        if (unlikely(SIM_attr_data_size(val) != len)) {
+            SIM_c_attribute_error(
+                "Invalid serialized value of byte array: expected data of %u "
+                "bytes, got %u", len, SIM_attr_data_size(val));
+            return Sim_Set_Illegal_Value;
+        }
+        memcpy(data, SIM_attr_data(val), len);
+        return Sim_Set_Ok;
+    } else if (unlikely(!SIM_attr_is_list(val))) {
+        if (data_allowed) {
+            SIM_attribute_error("Invalid serialized representation of byte "
+                                "array: not a list or data");
+        } else {
+            SIM_attribute_error("Invalid serialized representation of array: "
+                                "not a list");
+        }
+        return Sim_Set_Illegal_Type;
+    } else if (unlikely(SIM_attr_list_size(val) != len)) {
+        SIM_c_attribute_error(
+            "Invalid serialized representation of %sarray: expected list of "
+            "%u elements, got %u", data_allowed ? "byte " : "", len,
+            SIM_attr_list_size(val));
+        return Sim_Set_Illegal_Type;
+    }
+    attr_value_t *items = SIM_attr_list(val);
+    size_t children_elem_count = total_elem_count/len;
+    for (uint32 i = 0; i < len; ++i) {
+        set_error_t error;
+        if (dims == 1) {
+            error = deserialize_elem(&items[i], &data[i*elem_size]);
+        } else {
+            error = _deserialize_array_aux(
+                items[i], &data[i*children_elem_count*elem_size], elem_size,
+                dimsizes + 1, dims - 1, children_elem_count, deserialize_elem,
+                elems_are_bytes);
+        }
+        if (error != Sim_Set_Ok) {
+            return error;
+        }
+    }
+    return Sim_Set_Ok;
+}
+
+UNUSED static set_error_t
+_deserialize_array(attr_value_t in_val, void *out_data, size_t elem_size,
+                   const uint32 *dimsizes, uint32 dims,
+                   _deserializer_t deserialize_elem, bool elems_are_bytes) {
```

Comment:
There's a discrepancy between `_serialize_array` and `_deserialize_array` in that `elems_are_bytes` is signaled to `_serialize_array` by providing `NULL` as the `serialize_elem` function pointer. For `_deserialize_array`, it needs to be its own dedicated parameter, as in that case `_deserialize_array` must be flexible in whether it receives an integer list or data value as the final dimension, so it needs `deserialize_elem`.

Is this discrepancy ugly enough to change `_serialize_array` so that it also has a dedicated parameter for `elems_are_bytes`?

Reply:
I think its fine.